File: bin/unpack/simput-unpack.js
#! /usr/bin/env node

var shell = require('shelljs');

var path = require('path');
var fs = require('fs');

function getPath(arg) {
  return path.normalize(path.isAbsolute(arg) ? arg : path.join(process.cwd(), arg));
}

if (process.argv.length < 4) {
  console.log('unpack file.json output/directory');
} else {
  var filePath = getPath(process.argv[2]);
  var destPath = getPath(process.argv[3]);
  var jsonContent = require(filePath);
  var content = jsonContent.results || jsonContent;

  /* Results */

  Object.keys(content).forEach(function(name) {
    var fileDest = path.join(destPath, name);
    var destDir = path.dirname(fileDest);
    shell.mkdir('-p', destDir);
    fs.writeFileSync(fileDest, content[name]);
  });

  /* Copies */

  if (jsonContent.copies) {
    jsonContent.copies.forEach(function(filePath) {
      var sourceFile = path.join(process.cwd(), 'types', jsonContent.model.data.type, 'src', 'templates', filePath);
      var destinationFile = path.join(destPath, filePath);

      shell.mkdir('-p', path.dirname(destinationFile));
      shell.cp(sourceFile, destinationFile);
    });
  }

  /* Permissions */

  if (jsonContent.permissions) {
    var permissions = jsonContent.permissions;
    Object.keys(permissions).forEach(function(name) {
      var fileDest = path.join(destPath, name);
      shell.chmod(permissions[fileDest], fileDest);
    });
  }
}

File: bin/unpack/simput-unpack.py
#! /usr/bin/python

import json, os, sys, errno, stat, shutil

def getPath(value):
  if os.path.isabs(value):
    return os.path.normpath(value)
  return os.path.normpath(os.path.join(os.getcwd(), value))

def mkdirp(value):
  try:
    os.makedirs(value)
  except OSError as exc:
    if exc.errno == errno.EEXIST and os.path.isdir(value):
        pass
    else:
        raise

def addExecFlag(value):
  st = os.stat(value)
  os.chmod(value, st.st_mode | stat.S_IEXEC)

def unpack(srcFile, destDirectory):
  with open(srcFile) as simputExport:
    data = json.load(simputExport)
    fileContents = data['results'] if 'results' in data else data
    copies = data['copies'] if 'copies' in data else []
    permissions = data['permissions'] if 'permissions' in data else {}

    # Files

    for name in fileContents:
      fileDest = os.path.join(destDirectory, name)
      destDir = os.path.dirname(fileDest)
      mkdirp(destDir)
      with open(fileDest, 'w+') as out:
        out.write(fileContents[name])

    # Copies

    for entry in copies:
      sourceFile = entry['src']
      destinationFile = os.path.join(destDirectory, entry['dst']);

      mkdirp(os.path.dirname(destinationFile))
      shutil.copyfile(sourceFile, destinationFile)

    # Permissions

    for name in permissions:
      fileDest = os.path.join(destDirectory, name)
      # FIXME we assume the permission edits are to add exec flag
      addExecFlag(fileDest)

if __name__ == "__main__":
  if len(sys.argv) < 3:
    print 'unpack file.json output/directory'
  else:
    unpack(getPath(sys.argv[1]), getPath(sys.argv[2]))
